### PR TITLE
app-cdr/ecm: new package

### PR DESCRIPTION
**Pull request by yegortimoshenko**
ECM is a popular tool used to compress and decompress CD images. CDs have an error correction mechanism built in, and ECM strips that additional data, giving 15%+ in data savings. Stripped data can be recovered by using `unecm` utility.

This package in other distros/package managers:

* [Homebrew](https://github.com/Homebrew/homebrew-core/blob/36387cce5baa181fb99528ecf346ded3ef5a4967/Formula/ecm.rb) (macOS)
* [Arch Linux](https://www.archlinux.org/packages/community/i686/ecm-tools/)
* [FreeBSD](https://svnweb.freebsd.org/ports/head/archivers/ecm/)
* [Debian](https://packages.debian.org/jessie/ecm)
* [Google Play](https://play.google.com/store/apps/details?id=fr.bobenrieth.unecm&hl=en) (10K-50K installs)

Gentoo-Bug: https://bugs.gentoo.org/show_bug.cgi?id=618468

### CODE REVIEW

**Comment by yegortimoshenko** [comment]
On second thought, it's a much better fit at app-cdr category rather than app-arch (and it's under app-cdr in many overlays).

**Comment by mgorny** [comment]
Just to be clear, we're talking of complete image formats and not plain .iso dumps?

File: app-cdr/ecm/ecm-1.0.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+LICENSE="GPL-2+"
+SLOT="0"
+KEYWORDS="~amd64 ~x86 ~arm ~mips ~ppc ~ppc64 ~amd64-linux ~x86-linux ~x64-macos"
+
+DEPEND="app-arch/unzip"
+
+S="${WORKDIR}"
+
+src_compile() {
+	for i in *.c; do
```
Comment (mgorny) [inline app-cdr/ecm/ecm-1.0.0.ebuild]: Not a good solution. This makes parallel make inefficient, and as a result breaks e.g. distcc. I get this package probably ain't very large but still not a nice way of doing it. You can probably play with inline make, e.g. `emake -f - <<-EOF` (as a cheap alternative to adding files) ;-).

File: app-cdr/ecm/metadata.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+<maintainer type="person">
+    <email>yegortimoshenko@gmail.com</email>
+    <name>Yegor Timoshenko</name>
+</maintainer>
+</pkgmetadata>
```
Comment (mgorny) [inline app-cdr/ecm/metadata.xml]: You forgot to add us! ;-) Who will commit for you then?

**Comment by Q3CPMA** [comment]
Hello, this should be better. Yegor, I also added the doc use flag you were talking about.
```
# Copyright 1999-2017 Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2

EAPI="6"

inherit toolchain-funcs versionator

MY_PV=$(delete_all_version_separators)

DESCRIPTION="Convert between ECM and BIN files"
HOMEPAGE="https://web.archive.org/web/20091104002036/www.neillcorlett.com/${PN}"
SRC_URI="https://web.archive.org/web/20091021035854/www.neillcorlett.com/downloads/${PN}${MY_PV}.zip"

LICENSE="GPL-2+"
SLOT="0"
KEYWORDS="amd64 x86 ~arm ~mips ~ppc ~ppc64 ~amd64-linux ~x86-linux ~x64-macos"
IUSE="doc"

DEPEND="app-arch/unzip"

S="${WORKDIR}"

src_compile() {
        emake -f - <<-EOF
                all: ecm unecm
                %: %.c
        EOF
}

src_install() {
        dobin {un,}ecm
        if use doc; then
                dodoc {format,readme}.txt
        fi
}
```

File: app-cdr/ecm/ecm-1.0.0.ebuild (hunk 1 above)
Comment (yegortimoshenko) [inline app-cdr/ecm/ecm-1.0.0.ebuild]: Didn't know that make can read from stdin, cool, thank you!

File: app-cdr/ecm/metadata.xml (hunk 2 above)
Comment (yegortimoshenko) [inline app-cdr/ecm/metadata.xml]: Oops. I'll need to do that for all other packages as well.

**Comment by yegortimoshenko** [comment]
Hadrien, thanks! I thought it might need `doc` USE flag, but there is just not enough documentation files to justify it.

**Comment by yegortimoshenko** [comment]
@mgorny It's very generic, one can even strip a plain text file using `ecm` and then `unecm` and it will roundtrip successfully (without any compression though), so any format will work with `ecm`, not just ISO. Original author [describes](https://web.archive.org/web/20091105175002/http://www.neillcorlett.com:80/ecm) it as follows:

> ECM (Error Code Modeler) lets you prepare CD image files (such as BIN, CDI, NRG, CCD, or similar) so that they'll compress far better in programs such as WinRAR.

File: app-cdr/ecm/ecm-1.0.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+MY_PV=$(delete_all_version_separators)
+
+DESCRIPTION="CD EDC/ECC code (un)stripper"
+HOMEPAGE="https://web.archive.org/web/20091104002036/www.neillcorlett.com/ecm"
+SRC_URI="https://web.archive.org/web/20091021035854/www.neillcorlett.com/downloads/${PN}${MY_PV}.zip"
+
+LICENSE="GPL-2+"
+SLOT="0"
+KEYWORDS="~amd64 ~x86 ~arm ~mips ~ppc ~ppc64 ~amd64-linux ~x86-linux ~x64-macos"
```
Comment (mgorny) [inline app-cdr/ecm/ecm-1.0.0.ebuild]: Do you have Gentoo installs for all those systems? VMs count but qemu-user doesn't ;-).

File: app-cdr/ecm/ecm-1.0.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+src_compile() {
+	emake CC="$(tc-getCC)" -f - <<-'EOF'
+	      all: ecm unecm
+	EOF
+}
+
+src_test() {
+	einfo "roundtripping test.bin (single sector)"
+	zcat "${FILESDIR}/test.bin.gz" > "${T}/test.bin.orig" || die "failed to decompress"
```
Comment (mgorny) [inline app-cdr/ecm/ecm-1.0.0.ebuild]: You can't put binary files in FILESDIR. You can either use SRC_URI to fetch it from somewhere, or you need to encode it e.g. in base64.

File: app-cdr/ecm/ecm-1.0.0.ebuild (hunk 3 above)
Comment (yegortimoshenko) [inline app-cdr/ecm/ecm-1.0.0.ebuild]: I only have `amd64`, `x86`, `arm`, `~amd64-linux`, `~x86-linux` and `~x64-macos`.
So no `mips`, `ppc`, `ppc64`. I've inferred these arches from the fact that Debian builds this program for these arches: https://packages.debian.org/jessie/ecm

**Comment by Q3CPMA** [comment]
Hmmm, does the `CC="$(tc-getCC)"` serve a purpose? I thought make implicit rules automatically used `$(CC)`.

**Comment by yegortimoshenko** [comment]
@Q3CPMA See https://github.com/gentoo/gentoo/pull/4669#discussion_r117317191

**Comment by leio** [comment]
I am not interested of a ~mips keyword on this.

**Pull request closed without merging**